### ISSUE TITLE
v0.10.1.0.0 install error on macOS

### ISSUE THREAD

**Issue by mjsteinbaugh** [RcppCore/RcppArmadillo#314]
I'm seeing this error when attempting to install RcppArmadillo v0.10.1.0.0 on macOS:

```
> install.packages("RcppArmadillo", type = "source")
Installing package into ‘/usr/local/koopa/opt/r/4.0/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.rstudio.com/all/352/src/contrib/RcppArmadillo_0.10.1.0.0.tar.gz'
Content type 'binary/octet-stream' length 1642964 bytes (1.6 MB)
==================================================
downloaded 1.6 MB

* installing *source* package ‘RcppArmadillo’ ...
** package ‘RcppArmadillo’ successfully unpacked and MD5 sums checked
** using staged installation
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables...
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether clang++ -mmacosx-version-min=10.13 -std=gnu++11 accepts -g... yes
checking how to run the C++ preprocessor... clang++ -mmacosx-version-min=10.13 -std=gnu++11 -E
checking whether we are using the GNU C++ compiler... (cached) yes
checking whether clang++ -mmacosx-version-min=10.13 -std=gnu++11 accepts -g... (cached) yes
checking whether we have a suitable tempdir... /var/folders/l1/8y8sjzmn15v49jgrqglghcfr0000gn/T
checking whether R CMD SHLIB can already compile programs using OpenMP... yes
checking LAPACK_LIBS... R-supplied partial LAPACK found
configure: WARNING: Some complex-valued LAPACK functions may not be available
configure: creating ./config.status
config.status: creating inst/include/RcppArmadilloConfigGenerated.h
config.status: creating src/Makevars
** libs
clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG  -I'/usr/local/koopa/opt/r/4.0/site-library/Rcpp/include' -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include  -I../inst/include  -fPIC  -Wall -g -O2  -c RcppArmadillo.cpp -o RcppArmadillo.o
In file included from RcppArmadillo.cpp:22:
In file included from ../inst/include/RcppArmadillo.h:31:
In file included from ../inst/include/RcppArmadilloForward.h:49:
In file included from /usr/local/include/armadillo:621:
In file included from /usr/local/include/armadillo_bits/Mat_meat.hpp:9921:
../inst/include/RcppArmadillo/Mat_meat.h:104:11: error: member initializer 'n_alloc' does not name a non-static data member or
      base class
        , n_alloc(0)
          ^~~~~~~~~~
../inst/include/RcppArmadillo/Mat_meat.h:124:11: error: member initializer 'n_alloc' does not name a non-static data member or
      base class
        , n_alloc(0)
          ^~~~~~~~~~
2 errors generated.
make: *** [RcppArmadillo.o] Error 1
ERROR: compilation failed for package ‘RcppArmadillo’
* removing ‘/usr/local/koopa/opt/r/4.0/site-library/RcppArmadillo’
* restoring previous ‘/usr/local/koopa/opt/r/4.0/site-library/RcppArmadillo’
Warning in install.packages("RcppArmadillo", type = "source") :
  installation of package ‘RcppArmadillo’ had non-zero exit status
```

**Comment by eddelbuettel** [comment]
Is there anything special about your machine or compiler?  We have not seen this, and CRAN has macOS binaries:

https://cloud.r-project.org/web/packages/RcppArmadillo/index.html

**Comment by mjsteinbaugh** [comment]
Hi @eddelbuettel, I'm looking to see if there's anything weird with my config. The previous version 0.9.900.3.0 installs without issue:

```
install.packages("RcppArmadillo", type = "source")
Installing package into ‘/usr/local/koopa/opt/r/4.0/site-library’
(as ‘lib’ is unspecified)
trying URL 'https://packagemanager.rstudio.com/all/349/src/contrib/RcppArmadillo_0.9.900.3.0.tar.gz'
```

I'm using Homebrew and this [Rprofile.site](https://github.com/acidgenomics/koopa/blob/master/etc/R/Rprofile.site) configuration.

**Comment by mjsteinbaugh** [comment]
And here's my session info:

```
> sessioninfo::session_info()
─ Session info ───────────────────────────────────────────────────────────────
 setting  value
 version  R version 4.0.3 (2020-10-10)
 os       macOS Catalina 10.15.7
 system   x86_64, darwin17.0
 ui       X11
 language (EN)
 collate  en_US.UTF-8
 ctype    en_US.UTF-8
 tz       America/New_York
 date     2020-10-23

─ Packages ───────────────────────────────────────────────────────────────────
 package     * version date       lib source
 assertthat    0.2.1   2019-03-21 [1] CRAN (R 4.0.0)
 bb8         * 0.2.31  2020-10-23 [1] local
 BiocManager   1.30.10 2019-11-16 [1] CRAN (R 4.0.0)
 cli           2.1.0   2020-10-12 [1] RSPM (R 4.0.3)
 crayon        1.3.4   2017-09-16 [1] CRAN (R 4.0.0)
 fansi         0.4.1   2020-01-08 [1] CRAN (R 4.0.0)
 glue          1.4.2   2020-08-27 [1] RSPM (R 4.0.2)
 magrittr      1.5     2014-11-22 [1] CRAN (R 4.0.0)
 sessioninfo   1.1.1   2018-11-05 [1] CRAN (R 4.0.0)
 withr         2.3.0   2020-09-22 [1] RSPM (R 4.0.2)

[1] /usr/local/koopa/opt/r/4.0/site-library
[2] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
>
```

**Comment by eddelbuettel** [comment]
Maybe you should consult with the r-sig-mac list, I know little here ... other than that CRAN does AFAIK *not* use homebrew.

**Comment by mjsteinbaugh** [comment]
OK I can post a question there. I'm using the Homebrew R cask, which installs the official CRAN R framework.

**Comment by eddelbuettel** [comment]
I have a suspicion.  There was a one-line change there to accomodate Armadillo 10.1.0 -- could it be possible that you might have an older Armadillo header library "somewhere" that gets included instead / as-well or earlier?

**Comment by mjsteinbaugh** [comment]
Could this be due to the recent update to LLVM 11?

```
❯ gcc --version
Configured with: --prefix=/Library/Developer/CommandLineTools/usr --with-gxx-include-dir=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/c++/4.2.1
Apple clang version 11.0.3 (clang-1103.0.32.62)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Library/Developer/CommandLineTools/usr/bin
```

I see in the [CRAN checks](https://cloud.r-project.org/web/checks/check_flavors.html#r-release-macos-x86_64) that the package is tested against LLVM 10: Apple LLVM version 10.0.0 (clang-1000.10.44.4); GNU Fortran (GCC) 8.2.0

**Comment by mjsteinbaugh** [comment]
I'll double check the Armadillo library setup on my machine. Thanks

**Comment by eddelbuettel** [comment]
I don't know. I only have `g++` and `clang++` up to 10 on my box.  That initialisation of the `n_alloc(0)` is however very very vanilla C++ so it should not make it go haywire.

**Comment by mjsteinbaugh** [comment]
If any Mac users hit this issue, here's my armadillo config:

```sh
brew info armadillo
```

```
armadillo: stable 9.900.3 (bottled)
C++ linear algebra library
https://arma.sourceforge.io/
/usr/local/Cellar/armadillo/9.900.3 (597 files, 5MB) *
  Poured from bottle on 2020-10-23 at 15:58:08
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/armadillo.rb
License: Apache-2.0
==> Dependencies
Build: cmake ✔, pkg-config ✔
Required: arpack ✔, hdf5 ✔, openblas ✔, superlu ✔, szip ✔
```

**Comment by eddelbuettel** [comment]
So try uninstalling  that `brew` package and see if `RcppArmadillo` then installs -- it might.  And if it does and if this were to prove that the compiler goes looking in the wrong places ... then we'd also know we over here have no part in that crime :)

**Comment by mjsteinbaugh** [comment]
Updating the armadillo Homebrew recipe to 10.1.0 appears to fix this issue:

```
/usr/local/Cellar/armadillo
/usr/local/Cellar/armadillo/10.1.0/include/armadillo
/usr/local/Cellar/armadillo/10.1.0/include/armadillo_bits
/usr/local/Cellar/armadillo/10.1.0/lib/libarmadillo.10.1.0.dylib
/usr/local/Cellar/armadillo/10.1.0/lib/libarmadillo.10.dylib
/usr/local/Cellar/armadillo/10.1.0/lib/libarmadillo.dylib
/usr/local/Cellar/armadillo/10.1.0/lib/pkgconfig/armadillo.pc
/usr/local/Cellar/armadillo/10.1.0/share/Armadillo
/usr/local/Cellar/armadillo/10.1.0/share/Armadillo/CMake/ArmadilloConfig.cmake
/usr/local/Cellar/armadillo/10.1.0/share/Armadillo/CMake/ArmadilloConfigVersion.cmake
/usr/local/Cellar/armadillo/10.1.0/share/Armadillo/CMake/ArmadilloLibraryDepends-release.cmake
/usr/local/Cellar/armadillo/10.1.0/share/Armadillo/CMake/ArmadilloLibraryDepends.cmake
/usr/local/Cellar/cmake/3.18.4/share/cmake/Help/module/FindArmadillo.rst
/usr/local/Cellar/cmake/3.18.4/share/cmake/Modules/FindArmadillo.cmake
/usr/local/Cellar/cmake/3.18.4/share/doc/cmake/html/_sources/module/FindArmadillo.rst.txt
/usr/local/Cellar/cmake/3.18.4/share/doc/cmake/html/module/FindArmadillo.html
/usr/local/include/armadillo
/usr/local/include/armadillo_bits
/usr/local/lib/libarmadillo.10.1.0.dylib
/usr/local/lib/libarmadillo.10.dylib
/usr/local/lib/libarmadillo.dylib
/usr/local/lib/pkgconfig/armadillo.pc
/usr/local/opt/armadillo
/usr/local/share/Armadillo
/usr/local/share/cmake/Modules/FindArmadillo.cmake
/usr/local/var/homebrew/linked/armadillo
```

I'm working on a pull request for Homebrew. Thanks!

**Comment by eddelbuettel** [comment]
Good to know --- That is a serious issue though.  The R include instruction (to the headers shipped by RcppArmadillo) really *really* need to come first and that ordering is usually respected.  That issue is something you may want to take to r-sig-mac though.

**Comment by conradsnicta** [comment]
@eddelbuettel The actual bug seems to be in RcppArmadillo itself.  Specifically this innocuous looking line in RcppArmadilloForward.h:
https://github.com/RcppCore/RcppArmadillo/blob/master/inst/include/RcppArmadilloForward.h#L49

    #include <armadillo>

Note the use of angle brackets: `<...>`
In practice there is a difference between the use of quotes instead of angle brackets in the include statement.
In other words, there is a difference between  `#include <...>` and `#include "..."`

Quoting from https://stackoverflow.com/questions/21593/what-is-the-difference-between-include-filename-and-include-filename

> For #include "filename" the preprocessor searches first in the same directory as the file containing the directive, and then follows the search path used for the #include <filename> form. This method is normally used to include programmer-defined header files.

@mjsteinbaugh has two installations of Armadillo, and RcppArmadillo is getting confused which one to use. One installation is via Homebrew, which places it in /usr/local/include/, and the other installation is within RcppArmadillo itself.

Because of the two installations, the `#include <armadilo>` line in RcppArmadilloForward.h causes the compiler to get confused.  I suspect that changing the line to  `#include "armadilo"` would solve the issue.

**Comment by eddelbuettel** [comment]
Maybe, maybe not.  These include lines are also _ten years_ old and have hence survived a bazillion compilations.

**Comment by conradsnicta** [comment]
People have experienced this issue many times, and this is the most likely cause.

Related pull request: https://github.com/RcppCore/RcppArmadillo/pull/315
This won't break anything, and is likely to improve the situation.

**Comment by eddelbuettel** [comment]
> People have experienced this issue many times

Evidence?

**Comment by conradsnicta** [comment]
> > People have experienced this issue many times
> 
> Evidence?

See:
* https://github.com/RcppCore/RcppArmadillo/pull/281
* https://github.com/RcppCore/RcppArmadillo/issues/280
* https://github.com/RcppCore/RcppArmadillo/issues/264

From memory there's probably also several comments and/ir questions on Stackoverflow, but I don't have the time to look for them.

**Comment by eddelbuettel** [comment]
It's late here, I have not interest in arguing over this but it seems that a) those three issues are really one, b) they have already been discussed, c) you haven't actually demonstrated your suggestion fixes it (so you could have waited on @mjsteinbaugh to test it) nor do we know of side effects.

**Comment by conradsnicta** [comment]
They have been discussed, but the underlying issue (root cause) still stands: the compiler is getting confused which armadillo header to include.

All the other solutions in effect paper over the root cause, without addressing it directly.

**Comment by eddelbuettel** [comment]
I am locking this now.

**Comment by eddelbuettel** [comment]
@mjsteinbaugh Would you be able to test this?  

_I.e._ revert to a `brew` installed Armadillo version 9.* and then try installing RcppArmadillo with adjusted `#include` statements?

**Comment by kevinushey** [comment]
FWIW, a simple way to reproduce on macOS:

```
cd /tmp
echo '#error "You included the wrong version of RcppArmadillo!"' > /usr/local/include/armadillo
git clone https://github.com/RcppCore/RcppArmadillo
R CMD INSTALL --preclean RcppArmadillo
```

You'll then see:

```
clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG  -I'/Users/kevinushey/Library/R/4.0/library/Rcpp/include' -I/usr/local/include  -I../inst/include  -fPIC  -Wall -g -O2  -c RcppArmadillo.cpp -o RcppArmadillo.o
In file included from RcppArmadillo.cpp:21:
In file included from ../inst/include/RcppArmadillo.h:31:
In file included from ../inst/include/RcppArmadilloForward.h:49:
/usr/local/include/armadillo:1:2: error: "You included the wrong version of RcppArmadillo!"
#error "You included the wrong version of RcppArmadillo!"
 ^
```

This happens on macOS because R's `Makeconf` defaults to putting `/usr/local/include` on the search path before the package's own includes, via `CPPFLAGS`:

```
$ R CMD config CPPFLAGS
-I/usr/local/include
```

This isn't the case with other typical build sof R, hence why the issue only pops up on macOS (or, with users who have modified `CPPFLAGS` in a similar way).

Ultimately, I think using `#include "armadillo"` is indeed the right fix to guard against this.

**Comment by eddelbuettel** [comment]
> this happens on macOS because R's Makeconf defaults to putting /usr/local/include on the search path

Do you know why? Is that something we should bubble up to Simon and R Core?

I had been thinking about way to reproduce it on Linux and couldn't think of any because the (per-compiler) docs on quotes versus brackets are pretty clear that the include path should be walked, and we usuall get `-I.,./inst/include` first.

**Comment by kevinushey** [comment]
I don't know for sure, but my guess is that because most "package managers" on macOS install libraries into `/usr/local/include` rather than `/usr/include`, this was the most straightforward way for R to ensure that installed packages could find those installed libraries.

It does feel like this should be a part of `CFLAGS` / `CXXFLAGS` rather than `CPPFLAGS`, so that it would come after rather than before package-specific includes, though. But I have no idea how receptive R Core would be to such a change...

**Comment by kevinushey** [comment]
Also, I'm seeing a separate issue when I try to install the GitHub version of RcppArmadillo on everyone's favorite platform, macOS...

```
Error: package or namespace load failed for ‘RcppArmadillo’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/kevinushey/Library/R/4.0/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so':
  dlopen(/Users/kevinushey/Library/R/4.0/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so, 6): Symbol not found: __ZN4arma12arma_version5majorE
  Referenced from: /Users/kevinushey/Library/R/4.0/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so
  Expected in: flat namespace
 in /Users/kevinushey/Library/R/4.0/library/00LOCK-RcppArmadillo/00new/RcppArmadillo/libs/RcppArmadillo.so
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/kevinushey/Library/R/4.0/library/RcppArmadillo’
* restoring previous ‘/Users/kevinushey/Library/R/4.0/library/RcppArmadillo’
```

This is with:

```
$ clang -v
Apple clang version 12.0.0 (clang-1200.0.32.2)
Target: x86_64-apple-darwin19.6.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

Unless anyone else is seeing this I'll leave it alone; otherwise I'll try to dig in later tonight.

**Comment by eddelbuettel** [comment]
Please dig if you have the time.  The GitHub version is very close to what is on CRAN (and you can see the details in the few short commits; it's just minor cleanups; one for everybody's other favourite platform, CentOS 7 ....)

**Comment by kevinushey** [comment]
It looks like it may have been caused by https://github.com/RcppCore/RcppArmadillo/commit/273542909ef229822ac87e4cc7c8f9d087c7b0d1#diff-ed52c80292e10740987ed049314f15280e3220cb380bd4452212d3b549cda9e5. If I revert that commit, then all is well.

I can reproduce the issue on Ubuntu 20.04 if I switch the compiler to `clang`; e.g. by setting `CXX11 = clang++` in `~/.R/Makevars`. For posterity:

```
$ clang++ -v
clang version 10.0.0-4ubuntu1
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Found candidate GCC installation: /usr/lib/gcc/x86_64-linux-gnu/9
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-linux-gnu/9
Candidate multilib: .;@m64
Selected multilib: .;@m64
```

So it seems like some kind of compiler bug / "disagreement" between gcc and clang?

To complicate things more, it looks like the installation fails with `clang++ -std=gnu++11`, but succeeds with `clang++ -std=gnu++17`. So this might actually be due to a change in the C++ standard.

Searching a bit, I found this:

https://stackoverflow.com/a/39653928/1342082

and so it seems like this really is a change in behavior in the C++ standards. Before, `static constexpr` variables required an explicit definition somewhere; now, they are implicitly `inline` and so do not require one.

---

tl;dr: I think https://github.com/RcppCore/RcppArmadillo/commit/273542909ef229822ac87e4cc7c8f9d087c7b0d1 should be reverted, since those variable declarations are "technically" required by the C++11 standard (at least, as implemented by `clang`).

**Comment by eddelbuettel** [comment]
About to head off to get some groceries but why do you think in https://github.com/RcppCore/RcppArmadillo/commit/273542909ef229822ac87e4cc7c8f9d087c7b0d1 these are needed (apart from the fact that their presence _did_ cause an issue just now with CentOS 7 and its old `g++-4.8`?  We only refer to the values in the `arma` namespace and do not need the local placeholders.  Or did I overlook something?

**Comment by kevinushey** [comment]
The short version is that in C++11, this:

https://github.com/RcppCore/RcppArmadillo/blob/914a25fe6f9adae43e45ba0f2f3e93535da190e2/inst/include/armadillo_bits/arma_version.hpp#L31-L33

is treated only as a declaration (despite looking like a definition), whereas in C++17 it's treated as both a declaration + (inline) definition.

So, strictly speaking in C++11, we're still required to provide a definition for those declarations somewhere; that's what having

```
const unsigned int arma::arma_version::major;	
const unsigned int arma::arma_version::minor;	
const unsigned int arma::arma_version::patch;	
```

does for us in `RcppArmadillo.cpp`.

**Comment by eddelbuettel** [comment]
Ah, got it. In any event that was a post-CRAN release change (i.e. 0.10.1.0.1, not 0.10.1.0.0 as on CRAN) so it cannot have been the issue that upset OP at the beginning of the tread.

**Comment by conradsnicta** [comment]
If it's easier for everybody, I can change `constexpr` back to `const` within arma_version.hpp. 
Specifically, back to this:

    static const unsigned int major = ARMA_VERSION_MAJOR;
    static const unsigned int minor = ARMA_VERSION_MINOR;
    static const unsigned int patch = ARMA_VERSION_PATCH;

This should resolve the snafu with gcc 4.8

**Comment by conradsnicta** [comment]
If it's easier for everybody, I can change `constexpr` back to `const` within arma_version.hpp. 
Specifically, back to this:

    static const unsigned int major = ARMA_VERSION_MAJOR;
    static const unsigned int minor = ARMA_VERSION_MINOR;
    static const unsigned int patch = ARMA_VERSION_PATCH;

This is in addition to reverting https://github.com/RcppCore/RcppArmadillo/commit/273542909ef229822ac87e4cc7c8f9d087c7b0d1.

This should resolve the problem both with clang and gcc 4.8 (see https://github.com/RcppCore/RcppArmadillo/issues/313)

**Comment by conradsnicta** [comment]
> I already made (but have't committed) a change to simply use your UGLY_OLD_MACROS (which I generally hate) to avoid all that.

That's a good and robust solution.  Let's stick to this.

In the new [Armadillo 10.1.1 patch release](http://sourceforge.net/projects/arma/files/armadillo-10.1.1.tar.xz), I kept to the `constexpr` form.  
In other words, arma_version.hpp has `static constexpr unsigned int major = ARMA_VERSION_MAJOR`, etc.
